src/compiler/parser.ts

```diff
@@ -7210,6 +7221,8 @@ namespace Parser {
 
     function tryReuseAmbientDeclaration(pos: number): Statement | undefined {
         return doInsideOfContext(NodeFlags.Ambient, () => {
+            // TODO(jakebailey): this is totally wrong; `parsingContext` is the result of ORing a bunch of `1 << ParsingContext.XYZ`.
+            // The enum should really be a bunch of flags.

```

Comment:
All of the other uses get this right, at least, so node reuse has feasibly only been broken for ambient declarations.

Reply:
You're right - what happens if you pass in `0b111111111`? Which tests break? My hope is that *something* tested node reuse in an ambient context.

Reply:
Passing in `0b111111111` won't change anything because it uses equality checks, so they'll all just end up being false and no node reuse will occur.

Reply:
> won't change anything because it uses equality checks, so they'll all just end up being false and no node reuse will occur.

But we have tests that ensure that some node reuse occurs, don't we?

Reply:
All fourslash server tests which edit their files are incremental, which covers things in some aspect, but the `incrementalParser.ts` unittest file doesn't appear to test ambient declarations too much.

Reply:
Can you add an incremental test with an ambient context where node reuse *should* happen, and which fails when you pass in `0b11111111`?

Reply:
In trying to make a test case for this, I've found that this can only be a problem if there's an ambient declaration _inside_ of something else, like:

```ts
{
    declare module "foo" {
        // ...
    }
}
```

This is technically illegal code, and this bug here means we never use any nodes from this tree.

Ambients are only being reused _now_ because they can only be a child of the SourceFile itself, and its ParsingContext happens to be `ParsingContext.SourceElements = 0`. Then, the top level variable has value `1 << 0` or `1`, which is `ParsingContext.BlockStatements` which just so happens to also be reusable, so it works.

Reply:
So, I could fix the bug, but it'd only fix broken trees that probably never happen. But it's still totally wrong and only works because of a fluke!

Reply:
Okay, I'm wrong, the "ambient in a block" _also_ works because of a fluke; you get parsingContext `0b11` which equals `ParsingContext.SwitchClauseStatements` and that is also reusable, so the code passes. 🙃